fendermint/testing/materializer/tests/docker.rs

```diff
@@ -0,0 +1,132 @@
+    docker::{DockerMaterializer, DockerMaterials},
+    manifest::Manifest,
+    testnet::Testnet,
+    validation::validate_manifest,
+    TestnetName,
+};
+use futures::Future;
+use lazy_static::lazy_static;
+
+lazy_static! {
+    static ref CI_PROFILE: bool = std::env::var("PROFILE").unwrap_or_default() == "ci";
+    static ref STARTUP_WAIT_SECS: u64 = if *CI_PROFILE { 20 } else { 15 };
+    static ref TEARDOWN_WAIT_SECS: u64 = 5;
+    static ref PRINT_LOGS_ON_ERROR: bool = *CI_PROFILE;
+}
+
+/// Want to keep the testnet artifacts in the `tests/testnets` directory.
+fn tests_dir() -> PathBuf {
+    let dir = current_dir().unwrap();
+    debug_assert!(
+        dir.ends_with("materializer"),
+        "expected the current directory to be the crate"
+    );
+    dir.join("tests")
+}
+
+/// Directory where we keep the docker-materializer related data files.
+fn test_data_dir() -> PathBuf {
+    tests_dir().join("docker-materializer-data")
+}
+
+/// Parse a manifest from the `tests/manifests` directory.
+fn read_manifest(file_name: &str) -> anyhow::Result<Manifest> {
+    let manifest = tests_dir().join("manifests").join(file_name);
+    let manifest = std::fs::read_to_string(&manifest).with_context(|| {
+        format!(
+            "failed to read manifest from {}",
+            manifest.to_string_lossy()
+        )
+    })?;
+    let manifest = serde_yaml::from_str(&manifest).context("failed to parse manifest")?;
+    Ok(manifest)
+}
+
+/// Parse a manifest file in the `manifests` directory, clean up any corresponding
+/// testnet resources, then materialize a testnet and run some tests.
+pub async fn with_testnet<F>(manifest_file_name: &str, f: F) -> anyhow::Result<()>
+where
+    // https://users.rust-lang.org/t/function-that-takes-a-closure-with-mutable-reference-that-returns-a-future/54324
+    F: for<'a> FnOnce(
+        &Manifest,
+        &mut DockerMaterializer,
+        &'a mut Testnet<DockerMaterials, DockerMaterializer>,
+    ) -> Pin<Box<dyn Future<Output = anyhow::Result<()>> + 'a>>,
+{
+    let testnet_name = TestnetName::new(
+        PathBuf::from(manifest_file_name)
+            .file_stem()
+            .expect("there is a file step")
+            .to_string_lossy()
+            .to_string(),
+    );
+
+    let manifest = read_manifest(manifest_file_name)?;
+
+    // First make sure it's a sound manifest.
+    validate_manifest(&testnet_name, &manifest)
+        .await
+        .context("failed to validate manifest")?;
+
+    let mut materializer = DockerMaterializer::new(&test_data_dir(), 0)?;
+
+    materializer
+        .remove(&testnet_name)
+        .await
+        .context("failed to remove testnet")?;
+
+    let mut testnet = Testnet::setup(&mut materializer, &testnet_name, &manifest)
+        .await
+        .context("failed to set up testnet")?;
+
+    // Allow time for things to consolidate and blocks to be created.
+    tokio::time::sleep(Duration::from_secs(*STARTUP_WAIT_SECS)).await;
```

Comment:
nit: Optimally we want to wait until all nodes are up, maybe we can poll here the RPC API on the cometBFT container until a block has been produced?

Reply:
Yeah I agree, this is very crude. I think the place to do it though is in the test, where we at least know what nodes should exist, and what interface is available to query. We can always assume CometBFT I suppose, but I'm not sure every test will always be able to progress, maybe some setups are deliberately not starting a node that would allow quorum. 

Should we do these utilities as we add tests? I'm sure patterns will emerge.

Reply:
Maybe I can add a loop to wait until some general API responds from CometBFT, not necessarily block production.